lldb/tools/lldb-dap/DAP.h

```diff
@@ -392,9 +391,59 @@ struct DAP {
 
   void SetThreadFormat(llvm::StringRef format);
 
-  InstructionBreakpoint *GetInstructionBreakpoint(const lldb::break_id_t bp_id);
+  template <typename BreakpointType>
+  BreakpointType *GetBreakpointFromStopReason(lldb::SBThread &thread) {
+    // Check to see if have hit the <BreakpointType> breakpoint and change the
+    // reason accordingly, but only do so if all breakpoints that were
+    // hit are of <BreakpointType>.
+    const auto num = thread.GetStopReasonDataCount();
+    BreakpointType *bp = nullptr;
+    for (size_t i = 0; i < num; i += 2) {
+      lldb::break_id_t bp_id = thread.GetStopReasonDataAtIndex(i);
+      // If any breakpoint is not the <BreakpointType>, then stop and
+      // report this as a normal breakpoint
+      bp = GetBreakpoint<BreakpointType>(bp_id);
+      if (bp == nullptr)
+        return nullptr;
+    }
+    return bp;
+  }
+
+  template <typename BreakpointType>
+  BreakpointType *GetBreakpoint(const lldb::break_id_t bp_id);
+
+  template <>
+  FunctionBreakpoint *
+  GetBreakpoint<FunctionBreakpoint>(const lldb::break_id_t bp_id) {
+    for (auto &bp : function_breakpoints) {
```

Comment:
Because we're duplicating code here three times, would it make more sense that the template controls which breakpoint collection to grab, and then we just have the following logic in every case?

So we'd make GetBreakpoint

```
template <typename BreakpointType>
BreakpointType *GetBreakpoint(const lldb::break_id_t bp_id) {
    for (auto &bp : GetBreakpoints<>) {
      if (bp.second.bp.GetID() == bp_id)
        return &bp.second;
    }
    return nullptr;
}
```

Where the new `GetBreakpoints` call just gets the correct collection?

Reply:
Also, @satyajanga does the underlying collection support `find()`? Can we call that instead of enumerating ourselves

Reply:
updated to use std:find_if

the types for the collections are different,  so further templatization is challenging. Let me know if I am mis-undersanding. 
exception_breakpoints ->   std::optional<std::vector<ExceptionBreakpoint>> 
function_breakpoints -> llvm::StringMap<FunctionBreakpoint> 
instruction_breakpoints -> llvm::DenseMap<lldb::addr_t, InstructionBreakpoint>

I am open to changing. just didn't see a way to control the return collection type for template function that returns the collections.

Reply:
Didn't realize they're all different. Let's not worry about it